### PR TITLE
🌙 Add GCP pubsub MQL resources

### PR DESCRIPTION
**Pull request by imilchev**
Added GCP pubsub resources:
- topics - `gcp.project.pubsub.topics`
- subscriptions - `gcp.project.pubsub.subscriptions`
- snapshots - `gcp.project.pubsub.snapshots`

### CODE REVIEW

File: resources/packs/gcp/gcp.lr [hunk 1]
```diff
@@ -47,6 +47,8 @@ gcp.project @defaults("name") {
   // GCP Compute Resources for the Project
   // compute() gcp.compute
   clusters() []gcp.project.cluster
+  // PubSub-related resources
```
Comment (tas50) [inline resources/packs/gcp/gcp.lr]: These feed docs and autocomplete, so we want to write them high-level in context
```suggestion
  // GCP GCP PubSub Resources
```

File: resources/packs/gcp/gcp.lr [hunk 2]
```diff
@@ -1040,4 +1042,103 @@ private gcp.project.cluster.nodepool.config.confidentialNodes @defaults("enabled
   id string
   // Whether to use confidential nodes
   enabled bool
-}
\ No newline at end of file
+}
+
+private gcp.project.pubsub {
```
Comment (tas50) [inline resources/packs/gcp/gcp.lr]: We need docs comments on all of these

Comment (imilchev) [reply resources/packs/gcp/gcp.lr]: done

File: resources/packs/gcp/gcp.lr (hunk 1 above)
Comment (imilchev) [reply resources/packs/gcp/gcp.lr]: done

File: resources/packs/gcp/gcp.lr [hunk 3]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
   enabled bool
-}
\ No newline at end of file
+}
+
+// GCP PubSub resources
+private gcp.project.pubsub {
+  // Project ID
+  projectId string
+  // The list of topics in the current project
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // List of topics in the current project
```

File: resources/packs/gcp/gcp.lr [hunk 4]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  topics() []gcp.project.pubsub.topic
+  // The list of subscriptions in the current project
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // List of subscriptions in the current project
```

File: resources/packs/gcp/gcp.lr [hunk 5]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  subscriptions() []gcp.project.pubsub.subscription
+  // The list of snapshots in the current project
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // List of snapshots in the current project
```

File: resources/packs/gcp/gcp.lr [hunk 6]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  snapshots() []gcp.project.pubsub.snapshot
+}
+
+// GCP PubSub topic
+private gcp.project.pubsub.topic @defaults("name") {
+  // Internal ID
+  id string
+  // Project ID
+  projectId string
+  // The name of the topic
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Topic name
```

File: resources/packs/gcp/gcp.lr [hunk 7]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  name string
+  // The topic configuration
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Topic configuration
```

File: resources/packs/gcp/gcp.lr [hunk 8]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  config() gcp.project.pubsub.topic.config
+}
+
+// GCP PubSub topic configuration
+private gcp.project.pubsub.topic.config @defaults("kmsKeyName messageStoragePolicy") {
+  // Internal ID
+  id string
+  // The labels associated with this topic
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Labels associated with this topic
```

File: resources/packs/gcp/gcp.lr [hunk 9]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  labels map[string]string
+  // The name of the Cloud KMS key to be used to protect access to messages published to this topic
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Cloud KMS key used to protect access to messages published to this topic
```

File: resources/packs/gcp/gcp.lr [hunk 10]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  kmsKeyName string
+  // The message storage policy
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Message storage policy
```

File: resources/packs/gcp/gcp.lr [hunk 11]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  messageStoragePolicy gcp.project.pubsub.topic.config.messagestoragepolicy
+}
+
+// GCP PubSub topic message storage policy
+private gcp.project.pubsub.topic.config.messagestoragepolicy @defaults("allowedPersistenceRegions"){
+  // Internal ID
+  id string
+  // The list of GCP regions where messages that are published to the topic may be persisted in storage
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // List of GCP regions where messages published to the topic can persist in storage
```

File: resources/packs/gcp/gcp.lr [hunk 12]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  allowedPersistenceRegions []string
+}
+
+// GCP PubSub subscription
+private gcp.project.pubsub.subscription @defaults("name") {
+  // Internal ID
+  id string
+  // Project ID
+  projectId string
+  // The name of the subscription
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Subscription name
```

File: resources/packs/gcp/gcp.lr [hunk 13]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  name string
+  // The subscription configuration
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Subscription configuration
```

File: resources/packs/gcp/gcp.lr [hunk 14]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  config() gcp.project.pubsub.subscription.config
+}
+
+// GCP PubSub subscription configuration
+private gcp.project.pubsub.subscription.config @defaults("topic") {
+  // Internal ID
+  id string
+  // The topic to which the subscription points
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Topic to which the subscription points
```

File: resources/packs/gcp/gcp.lr [hunk 15]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  topic gcp.project.pubsub.topic
+  // Configuration for subscriptions that operate in push mode
+  pushConfig gcp.project.pubsub.subscription.config.pushconfig
+  // The default maximum time after a subscriber receives a message before the subscriber should acknowledge the message
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Default maximum time a subscriber can take to acknowledge a message after receiving it
```

File: resources/packs/gcp/gcp.lr [hunk 16]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  ackDeadline time
+  // Whether to retain acknowledged messages
+  retainAckedMessages bool
+  // How long to retain messages in backlog from the time of publish
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // How long to retain messages in the backlog after they're published
```

File: resources/packs/gcp/gcp.lr [hunk 17]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  expirationPolicy time
+  // The labels associated with this subscription
+  labels map[string]string
+}
+
+// GCP PubSub Configuration for subscriptions that operate in push mode
+private gcp.project.pubsub.subscription.config.pushconfig {
+  // Internal ID
+  id string
+  // A URL locating the endpoint to which messages should be pushed
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // URL of the endpoint to which to push messages
```

File: resources/packs/gcp/gcp.lr [hunk 18]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  attributes map[string]string
+}
+
+// GCP PubSub snapshot
+private gcp.project.pubsub.snapshot @defaults("name") {
+  // Internal ID
+  id string
+  // Project ID
+  projectId string
+  // The name of the subscription
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // Subscription name
```

File: resources/packs/gcp/gcp.lr [hunk 19]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  name string
+  // The topic for which the snapshot is
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr:1141]: Hey @imilchev is this the original topic that this is a snapshot of?

File: resources/packs/gcp/gcp.lr [hunk 20]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  topic gcp.project.pubsub.topic
+  // The expiration for the snapshot
```
Comment (misterpantz) [inline resources/packs/gcp/gcp.lr]:
```suggestion
  // When the snapshot expires
```

**Review by misterpantz — COMMENTED** [review]
Hiya! Thanks so much for providing clear docs. I made suggestions for making some of them more succinct and easier to read.

File: resources/packs/gcp/gcp.lr (hunk 19 above)
Comment (imilchev) [reply resources/packs/gcp/gcp.lr:1141]: I believe so. All docs I took from the documentation of the Google API. I don't know all of those concepts myself either... 😞

File: resources/packs/gcp/gcp.lr [hunk 21]
```diff
@@ -794,7 +796,7 @@ private gcp.dns.policy {
   enableLogging bool
 }
 
-// GKE Cluster
+// GCP GKE Cluster
 private gcp.project.cluster @defaults("name") {
```
Comment (chris-rock) [inline resources/packs/gcp/gcp.lr:800]: @tas50 should we rename `gcp.project.cluster` to `gcp.project.gke`?

File: resources/packs/gcp/gcp.lr [hunk 22]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
   // Internal ID
   id string
   // Whether to use confidential nodes
   enabled bool
-}
\ No newline at end of file
+}
+
+// GCP PubSub resources
+private gcp.project.pubsub {
```
Comment (chris-rock) [inline resources/packs/gcp/gcp.lr]: This is conflicting with `gcp.project { pubsub() }`. We should rename it to something like `gcp.project.pubsubService`

File: resources/packs/gcp/gcp.lr [hunk 23]
```diff
@@ -1012,32 +1016,134 @@ private gcp.project.cluster.nodepool.config.kubeletConfig @defaults("cpuManagerP
...
+  labels map[string]string
+  // Cloud KMS key used to protect access to messages published to this topic
+  kmsKeyName string
+  // Message storage policy
+  messageStoragePolicy gcp.project.pubsub.topic.config.messagestoragepolicy
+}
+
+// GCP PubSub topic message storage policy
+private gcp.project.pubsub.topic.config.messagestoragepolicy @defaults("allowedPersistenceRegions"){
+  // Internal ID
```
Comment (chris-rock) [inline resources/packs/gcp/gcp.lr]: Internal ID sounds strange. Is it the topic id?

File: resources/packs/gcp/pubsub.go [hunk 24]
```diff
@@ -0,0 +1,347 @@
...
+	for {
+		t, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		mqlTopic, err := g.MotorRuntime.CreateResource("gcp.project.pubsub.topic",
+			"id", t.ID(),
```
Comment (chris-rock) [inline resources/packs/gcp/pubsub.go]: is the id is the same as the name, why do we need the id field then?

File: resources/packs/gcp/pubsub.go [hunk 25]
```diff
@@ -0,0 +1,347 @@
...
+
+	messageStoragePolicy, err := g.MotorRuntime.CreateResource("gcp.project.pubsub.topic.config.messagestoragepolicy",
+		"id", fmt.Sprintf("%s/config/messagestoragepolicy", t.ID()),
+		"allowedPersistenceRegions", core.StrSliceToInterface(cfg.MessageStoragePolicy.AllowedPersistenceRegions),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return g.MotorRuntime.CreateResource("gcp.project.pubsub.topic.config",
+		"id", fmt.Sprintf("%s/config", t.ID()),
```
Comment (chris-rock) [inline resources/packs/gcp/pubsub.go]: why not `topicID`? The id you use here should be implemented in line 51

```
func (g *mqlGcpProjectPubsubTopicConfig) id() (string, error) {
    topicID, err := g.TopicId()
    if err != nil { return "", err }
    return "gcp.project.pubsub.topic.config/" + topicID + /config
}
```

The implementation above assumes that there is only one config per topic

File: resources/packs/gcp/pubsub.go [hunk 26]
```diff
@@ -0,0 +1,347 @@
...
+		s, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		topic, err := g.MotorRuntime.CreateResource("gcp.project.pubsub.topic",
+			"id", s.Topic.ID(),
```
Comment (chris-rock) [inline resources/packs/gcp/pubsub.go:376]: I do not think we need id here if it is the same as the name

File: resources/packs/gcp/pubsub.go [hunk 27]
```diff
@@ -0,0 +1,347 @@
...
+	defer pubsubSvc.Close()
+
+	s := pubsubSvc.Subscription(name)
+	cfg, err := s.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	topic, err := g.MotorRuntime.CreateResource("gcp.project.pubsub.topic",
+		"id", cfg.Topic.ID(),
```
Comment (chris-rock) [inline resources/packs/gcp/pubsub.go]: I do not think we need id here if it is the same as name

File: resources/packs/gcp/pubsub.go [hunk 28]
```diff
@@ -0,0 +1,347 @@
...
+	for {
+		s, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		mqlSub, err := g.MotorRuntime.CreateResource("gcp.project.pubsub.subscription",
+			"id", s.ID(),
```
Comment (chris-rock) [inline resources/packs/gcp/pubsub.go]: I do not think we need the id here if it is the same as the name

File: resources/packs/gcp/pubsub.go [hunk 29]
```diff
@@ -0,0 +1,347 @@
...
+		return nil, err
+	}
+
+	return g.MotorRuntime.CreateResource("gcp.project.pubsub",
+		"projectId", projectId,
+	)
+}
+
+func (g *mqlGcpProjectPubsubTopic) id() (string, error) {
+	return g.Id()
```
Comment (chris-rock) [inline resources/packs/gcp/pubsub.go]: The id implementation do not need to have an id field. You can generate the id from any field.

**Review by chris-rock — COMMENTED** [review]
Great addition @imilchev I added a few comments to make the implementation even a bit easier since we can remove most of the id fields from the resources.

File: resources/packs/gcp/gcp.lr (hunk 21 above)
Comment (imilchev) [reply resources/packs/gcp/gcp.lr:800]: I guess the suggestion here is to access the clusters like the other resources I added - `gcp.project.gke.clusters`?

File: resources/packs/gcp/pubsub.go (hunk 25 above)
Comment (imilchev) [reply resources/packs/gcp/pubsub.go]: it is, but that breaks if there are topics with the same name in multiple projects. To make this fully unique, I need to add the project as well

File: resources/packs/gcp/gcp.lr (hunk 19 above)
Comment (misterpantz) [reply resources/packs/gcp/gcp.lr:1141]: Aw, please don't be 😞 ! What if we change this to "The original topic of which this is a snapshot" ? I think that would be more helpful.

File: resources/packs/gcp/gcp.lr (hunk 21 above)
Comment (chris-rock) [reply resources/packs/gcp/gcp.lr:800]: We change this in another PR since it is not related to this PR

**Review by chris-rock — APPROVED** [review]
Great work @imilchev, I updated the GCP pubsub go mod dependency and added some minor tweaks to the default value.

**Pull request merged**